CORE-8695 crash_tracker: implement crash report upload markers

pgellert opened the pull request:
We will use empty files to track if a certain crash file has been uploaded through our metrics reporter telemetry mechanism.

To mark a crash file as uploaded, we touch a new file with the same name + the ".uploaded" suffix. To check if a crash file has been uploaded, we check if a file with the same name + the ".uploaded" suffix exists. We also implement the clean up of any upload markers on disk that do not have an associated crash report.

Fixes https://redpandadata.atlassian.net/browse/CORE-8695

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes
* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

pgellert committed:
crash_tracker: implement upload markers

Implement marking crash reports as uploaded and checking if a crash
report has been uploaded.

This is implemented using empty files created on disk with the same name
as the crash report + a ".uploaded" suffix.

pgellert committed:
crash_tracker: clean up dangling upload markers

Clean up any upload markers on disk that do not have an associated crash
report. This cleanup is done after the count-based removal of crash
reports (see remove_old_crashfiles()).

dotnwat commented on src/v/crash_tracker/recorder.cc:
```diff
@@ -33,6 +33,16 @@ namespace crash_tracker {
+std::filesystem::path
+to_upload_marker_path(const std::filesystem::path& crash_report_path) {
+    return crash_report_path.parent_path()
+           / (crash_report_path.filename().string() + recorder::upload_marker_suffix);
```
is this different than `return path + suffix`?

pgellert replied:
No, I've simplified this now.

dotnwat commented on src/v/crash_tracker/recorder.cc:
```diff
@@ -212,6 +222,23 @@ void recorder::record_crash_exception(std::exception_ptr eptr) {
+
+    // Create an empty upload marker
+    const auto marker_path = to_upload_marker_path(file_path).string();
+    auto f = co_await ss::open_file_dma(marker_path, ss::open_flags::create);
```
nit: without `O_EXCL` i'd expect this to succeed even if the file exists. So you could probably drop the `is_uploaded()` checked, unless there is something i'm missing.

pgellert replied:
Yeah, fair point, that works too.

dotnwat commented on src/v/crash_tracker/recorder.cc:
```diff
@@ -105,9 +113,36 @@ ss::future<> recorder::remove_old_crashfiles() const {
+    }
+
+    for (const auto& entry :
+         std::filesystem::directory_iterator(crash_report_dir)) {
```
use `utils/directory_walker.h`?

pgellert replied:
I've changed to `utils/directory_walker.h` now and added a commit (https://github.com/redpanda-data/redpanda/pull/25130/commits/a0a52ed6cb64f950e409c0eca65f53c9d38c69ca) to use `utils/directory_walker.h` in `recorder::get_recorded_crashes` as well.

pgellert committed:
crash_tracker: use seastar IO for listing crashes

Use directory_walker::walk instead of
std::filesystem::directory_iterator for gathering the list of recorded
crashes. Since directory_walker::walk does not expose the last write
time of the directory entry, we now explicitly stat the file for this
information.

There is no behaviour change intended otherwise.

pgellert commented:
[force-push](https://github.com/redpanda-data/redpanda/compare/2a40dba6daae6dd7fc03acf2116d21dad6bb082c..9eca7522528b3627179e233c18533602c30ee6c3): address code review feedback

dotnwat commented on src/v/crash_tracker/recorder.cc:
```diff
@@ -214,17 +215,43 @@ void recorder::record_crash_exception(std::exception_ptr eptr) {
+  std::vector<recorder::recorded_crash>& result,
+  recorder::include_malformed_files incl_malformed) {
+    const auto path = (basedir / std::string_view{entry.name}).string();
+    if (!path.ends_with(crash_report_suffix)) {
```
if you want to be more defensive you can add

`if (entry.type != ss::directory_entry_type::regular || !path.ends_with..) {`

since the entry could be all kinds of things (directories, block devices, etc...) all of which except type `regular` which is what you are looking for.

pgellert replied:
Yeah fair, but that's not necessary since redpanda owns this directory, and we only create files here for now. We can add this if we ever introduce other objects. I've left the code as is for now.

dotnwat replied:
> since redpanda owns this directory

this is exactly why it's useful to be defensive. to guard against future mistakes.

dotnwat commented on src/v/crash_tracker/recorder.cc:
```diff
@@ -34,6 +34,15 @@ namespace crash_tracker {
+
+std::filesystem::path
+to_upload_marker_path(const std::filesystem::path& crash_report_path) {
+    return crash_report_path.string() + recorder::upload_marker_suffix;
```
nit: i don't think you need to convert the path into the string first. just `return path + suffix` https://godbolt.org/z/dvv9hM3nr

pgellert replied:
Interesting... This only works withs with `operator+=` not `operator+`. Eg this compiles:
```
std::filesystem::path
to_upload_marker_path(std::filesystem::path crash_report_path) {
    crash_report_path += recorder::upload_marker_suffix;
    return crash_report_path;
}
```
But this doesn't compile:
```
std::filesystem::path
to_upload_marker_path(const std::filesystem::path& crash_report_path) {
    return crash_report_path + recorder::upload_marker_suffix;
}
```

dotnwat replied:
ha, interesting. TIL

dotnwat reviewed (dismissed):
lgtm

pgellert commented:
[force-push](https://github.com/redpanda-data/redpanda/compare/9eca7522528b3627179e233c18533602c30ee6c3..3f15feffcce5c17a9ae7e83eeb49fa42cedb1bb6): rebase to dev to fix merge conflict in `recorder.cc` (no other changes)

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--45991fd3-6583-4b6f-ae86-590c747b29da-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62262">build#62262</a></summary>


| test_id                                                                                                                                                                                             | test_kind   | job_url                                                                                   | test_status   | passed   |
|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| gtest_raft_rpunit.gtest_raft_rpunit                                                                                                                                                                 | unit        | https://buildkite.com/redpanda/redpanda/builds/62262#01953c9d-e11c-4208-af41-a1b38e08bdba | FLAKY         | 1/2      |
| rptest.tests.datalake.datalake_dlq_test.DatalakeDLQTest.test_invalid_record_action_runtime_change.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.use_topic_property=True | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62262#01953cf9-5b7c-4516-8e64-cf34246ce916 | FLAKY         | 1/2      |
| storage_e2e_single_thread_rpunit.storage_e2e_single_thread_rpunit                                                                                                                                   | unit        | https://buildkite.com/redpanda/redpanda/builds/62262#01953c9d-e11c-417c-997e-772575a8354a | FLAKY         | 1/2      |
</details>
<!--/body-->
